README.md
QrsDetector [![crates.io](https://img.shields.io/crates/v/qrs_detector.svg)](https://crates.io/crates/qrs_detector) [![docs.rs](https://img.shields.io/docsrs/qrs_detector?labelColor=1C2C2E&color=C96329&logo=rust)](https://docs.rs/qrs_detector/latest/qrs_detector/)
===========

This crate provides a realtime ECG QRS detector.

The implementation is based on [this article](https://biomedical-engineering-online.biomedcentral.com/articles/10.1186/1475-925X-3-28).


CHANGELOG.md
Unreleased
==========

 * **breaking:** Replaced `typenum` with const generics.
 * `SamplingFrequency` internal field is no longer accessible.
 * `SamplingFrequencyExt` is now implemented for `f64`.
 * Added `Thresholds`, `QrsDetector::thresholds()`.
 * Added `alloc` feature to automatically allocate F buffers.
 * Added `SamplingFrequency::{samples_to_s, samples_to_ms, raw}` functions.

0.2.0
==========

 * **breaking:** Remove preprocessing. Users must derive their differentiated complex leads.
 * **breaking:** Internal buffer size to calculate `F` changed from 350ms to 300ms.

0.1.2
=====

 * Small internal improvements

0.1.1
=====
 * Dependency update

0.1.0
=====
 * Initial release

tests/integration_test.rs
use qrs_detector::sampling::*;
use qrs_detector::QrsDetector;

#[test]
fn test_simulated_signal() {
    let mut detector = QrsDetector::new::<216, 36>(720.sps());

    let mut detections = 0;
    let mut prev: Option<f32> = None;
    let mut prev2: Option<f32> = None;

    let data = include_str!("./data/aami3a.txt");

    let samples = data
        .split_terminator('\n')
        .map(|str| str.trim().parse::<f32>().unwrap())
        .collect::<Vec<_>>();

    for window in samples.windows(4) {
        // A slight moving average filtering
        let sum = window.iter().sum::<f32>();
        let avg = sum / 4.0;
        if let Some(p) = prev.replace(avg) {
            if let Some(p2) = prev2.replace(p) {
                if let Some(_) = detector.update((p2 - avg).abs()) {
                    detections += 1;
                }
            }
        }
    }

    assert_eq!(38, detections);
}

#[cfg(feature = "alloc")]
#[test]
fn test_simulated_signal_alloc() {
    let mut detector = QrsDetector::new_alloc(720.sps());

    let mut detections = 0;
    let mut prev: Option<f32> = None;
    let mut prev2: Option<f32> = None;

    let data = include_str!("./data/aami3a.txt");

    let samples = data
        .split_terminator('\n')
        .map(|str| str.trim().parse::<f32>().unwrap())
        .collect::<Vec<_>>();

    for window in samples.windows(4) {
        // A slight moving average filtering
        let sum = window.iter().sum::<f32>();
        let avg = sum / 4.0;
        if let Some(p) = prev.replace(avg) {
            if let Some(p2) = prev2.replace(p) {
                if let Some(_) = detector.update((p2 - avg).abs()) {
                    detections += 1;
                }
            }
        }
    }

    assert_eq!(38, detections);
}


src/sliding.rs
//! Sliding window

use core::marker::PhantomData;

pub struct SlidingWindow<T, C> {
    buffer: C,
    idx: usize,
    full: bool,
    _marker: PhantomData<T>,
}

impl<T: Default + Copy, const N: usize> Default for SlidingWindow<T, [T; N]> {
    fn default() -> Self {
        Self::new([T::default(); N])
    }
}

impl<T, C> SlidingWindow<T, C>
where
    T: Copy,
    C: AsRef<[T]> + AsMut<[T]>,
{
    pub fn new(buffer: C) -> Self {
        Self {
            buffer,
            idx: 0,
            full: false,
            _marker: PhantomData,
        }
    }

    #[allow(dead_code)]
    pub fn clear(&mut self) {
        self.idx = 0;
        self.full = false;
    }

    pub fn capacity(&self) -> usize {
        self.buffer.as_ref().len()
    }

    pub fn len(&self) -> usize {
        if self.full {
            self.capacity()
        } else {
            self.idx
        }
    }

    pub fn last(&self) -> Option<T> {
        if self.idx == 0 && !self.full {
            None
        } else {
            let idx = if self.idx == 0 {
                self.capacity()
            } else {
                self.idx
            };
            Some(self.buffer.as_ref()[idx - 1])
        }
    }

    #[allow(dead_code)]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    pub fn is_full(&self) -> bool {
        self.full
    }

    pub fn push(&mut self, sample: T) -> Option<T> {
        let buffer = self.buffer.as_mut();
        let old = self.full.then_some(buffer[self.idx]);

        buffer[self.idx] = sample;
        self.idx = (self.idx + 1) % buffer.len();
        if self.idx == 0 {
            self.full = true;
        }

        old
    }

    pub fn iter_unordered(&self) -> impl Iterator<Item = T> + Clone + '_ {
        (0..self.len()).map(|i| self.buffer.as_ref()[i])
    }
}


src/sampling.rs
//! Helpers for working with sampling frequencies and sample numbers.

#[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
pub struct SamplingFrequency(f32);

/// Extension functions for numeric types used to create [`SamplingFrequency`] values.
///
/// # Usage
/// ```rust
/// use qrs_detector::sampling::*;
///
/// // Both values represent 500 samples per second
/// let fs = 500.sps();
/// let fs2 = 0.5.ksps();
///
/// assert_eq!(fs, fs2);
/// ```
pub trait SamplingFrequencyExt {
    fn sps(self) -> SamplingFrequency;
    fn ksps(self) -> SamplingFrequency;
}

impl SamplingFrequencyExt for f32 {
    fn sps(self) -> SamplingFrequency {
        SamplingFrequency(self)
    }

    fn ksps(self) -> SamplingFrequency {
        (self * 1000.0).sps()
    }
}

impl SamplingFrequencyExt for usize {
    fn sps(self) -> SamplingFrequency {
        SamplingFrequency(self as f32)
    }

    fn ksps(self) -> SamplingFrequency {
        (self * 1000).sps()
    }
}

impl SamplingFrequencyExt for f64 {
    fn sps(self) -> SamplingFrequency {
        SamplingFrequency(self as f32)
    }

    fn ksps(self) -> SamplingFrequency {
        (self * 1000.0).sps()
    }
}

impl SamplingFrequency {
    /// Returns the sampling frequency in units of samples per second.
    pub fn raw(self) -> f32 {
        self.0
    }

    /// Convert `ms` milliseconds to number of samples
    /// ```rust
    /// # use qrs_detector::sampling::*;
    /// #
    /// let samples = 500.sps().s_to_samples(2.5);
    /// assert_eq!(samples, 1250);
    /// ```
    pub fn ms_to_samples(self, ms: f32) -> usize {
        ((ms * self.0) as usize) / 1000
    }

    /// Convert `s` seconds to number of samples
    /// ```rust
    /// # use qrs_detector::sampling::*;
    /// #
    /// let samples = 500.sps().ms_to_samples(500.0);
    /// assert_eq!(samples, 250);
    /// ```
    pub fn s_to_samples(self, s: f32) -> usize {
        self.ms_to_samples(s * 1000.0)
    }

    /// Convert `samples` to seconds
    /// ```rust
    /// # use qrs_detector::sampling::*;
    /// #
    /// let s = 500.sps().samples_to_s(250);
    /// assert_eq!(s, 0.5);
    /// ```
    pub fn samples_to_s(self, samples: usize) -> f32 {
        (samples as f32) / self.0
    }

    /// Convert `samples` to milliseconds
    /// ```rust
    /// # use qrs_detector::sampling::*;
    /// #
    /// let s = 500.sps().samples_to_ms(250);
    /// assert_eq!(s, 500.0);
    /// ```
    pub fn samples_to_ms(self, samples: usize) -> f32 {
        self.samples_to_s(samples) * 1000.0
    }
}


src/algorithms/f.rs
use crate::{sampling::*, sliding::SlidingWindow};

#[derive(Copy, Clone, Debug)]
enum FState {
    Ignore(usize),
    Init(usize, f32),
    Integrate(f32),
}

pub struct F<FMW, FB> {
    fs: SamplingFrequency,

    /// F should be initialized at the same time as M is, skip earlier samples
    state: FState,

    /// 300ms of the individual max samples of the 50ms buffer.
    /// We store the maxima to save some memory and computation.
    f_max_window: SlidingWindow<f32, FMW>,

    /// 50ms window of the signal
    f_buffer: SlidingWindow<f32, FB>,
}

impl<FMW, FB> F<FMW, FB>
where
    FMW: AsRef<[f32]> + AsMut<[f32]>,
    FB: AsRef<[f32]> + AsMut<[f32]>,
{
    pub fn new(
        fs: SamplingFrequency,
        f_max_window: SlidingWindow<f32, FMW>,
        f_buffer: SlidingWindow<f32, FB>,
    ) -> Self {
        // sanity check buffer sizes
        debug_assert_eq!(
            f_max_window.capacity(),
            fs.ms_to_samples(300.0),
            "Incorrect FMW type parameter. Buffer must be {} samples long.",
            fs.ms_to_samples(300.0)
        );
        debug_assert_eq!(
            f_buffer.capacity(),
            fs.ms_to_samples(50.0),
            "Incorrect FB type parameter. Buffer must be {} samples long.",
            fs.ms_to_samples(50.0)
        );

        Self {
            fs,
            state: FState::Ignore(fs.s_to_samples(2.65)),
            f_max_window,
            f_buffer,
        }
    }

    pub fn clear(&mut self) {
        self.state = FState::Ignore(self.fs.s_to_samples(2.65));
        self.f_max_window.clear();
        self.f_buffer.clear();
    }

    fn update_f_buffers(&mut self, sample: f32) -> (Option<f32>, f32) {
        // TODO: there are some special cases where the max search can be skipped
        self.f_buffer.push(sample);

        // Calculate maximum value in the latest 50ms window
        let max: f32 = self
            .f_buffer
            .iter_unordered()
            .fold(0.0, |acc, x| acc.max(x));

        // Keep the 50ms maximum values for each sample in latest 300ms window
        // The oldest sample corresponds to the oldest 50ms in the latest 350ms window
        // TODO FIXME: off by some error :)
        let old = self.f_max_window.push(max);

        (old, max)
    }

    pub fn update(&mut self, sample: f32) {
        self.state = match self.state {
            FState::Ignore(1) => FState::Init(self.f_max_window.capacity() - 1, 0.0),
            FState::Ignore(n) => FState::Ignore(n - 1),
            FState::Init(n, favg) => {
                let favg = favg + sample;
                self.update_f_buffers(sample);

                if n == 0 {
                    FState::Integrate(favg.max(0.0) / (self.f_max_window.capacity() as f32))
                } else {
                    FState::Init(n - 1, favg)
                }
            }
            FState::Integrate(f) => {
                let (oldest_max, max) = self.update_f_buffers(sample);
                FState::Integrate((f + (max - oldest_max.unwrap()) / 150.0).max(0.0))
            }
        };
    }

    pub fn threshold(&self) -> Option<f32> {
        if let FState::Integrate(f) = self.state {
            Some(f)
        } else {
            None
        }
    }
}


src/algorithms/m.rs
use crate::{sampling::*, sliding::SlidingWindow};

#[derive(Copy, Clone, Debug)]
enum MState {
    Init(usize, f32),
    Disallow(usize, f32),
    Decreasing(usize, f32, f32),
    ConstantLow(f32),
}

pub struct M {
    state: MState,
    mm: SlidingWindow<f32, [f32; 5]>,
    fs: SamplingFrequency,
    pub current_decrement: f32,
}

impl M {
    pub fn new(fs: SamplingFrequency) -> Self {
        Self {
            fs,
            mm: SlidingWindow::default(),
            // Initially M = 0.6*max(Y) is set for the first 3 s [originally 5s] of the signal
            state: MState::Init(fs.s_to_samples(3.0), 0.0),
            current_decrement: 0.0,
        }
    }

    pub fn clear(&mut self) {
        self.mm.clear();
        self.state = MState::Init(self.fs.s_to_samples(3.0), 0.0);
        self.current_decrement = 0.0;
    }

    fn m(&self) -> f32 {
        // M is calculated as an average value of MM.
        // Divide by 5 was done while calculating the individual Mx values
        self.mm.iter_unordered().sum()
    }

    pub fn update(&mut self, sample: f32) {
        self.state = match self.state {
            MState::Init(0, m) => {
                // Initially M = 0.6*max(Y) is set for the first 3 s [originally 5s] of the signal
                // A buffer with 5 steep-slope threshold values is preset:
                // MM = [M1 M2 M3 M4 M5],
                // where M1 ÷ M5 are equal to M
                let m = 0.6 * m.max(sample);

                for _ in 0..5 {
                    self.mm.push(m / 5.0);
                }

                // It is not clear in the article what to do initially:
                // - wait for a QRS detection and then start the M algorithm with Disallow state
                // - decrease using "low slope" immediately
                // This implementation uses the second option

                // M is decreased in an interval 225 to 1225 ms [originally 200 to 1200 ms]
                // following the last QRS detection at a low slope, reaching 60 % of its
                // refreshed value at 1225 ms [originally 1200 ms].
                let n_samples = self.fs.s_to_samples(1.0);
                let decrement = m * 0.4 / n_samples as f32;
                self.current_decrement = decrement;

                MState::Decreasing(n_samples, m, decrement)
            }

            // Initially M = 0.6*max(Y) is set for the first 3 s [originally 5s] of the signal
            // Collect maximum value while in Init state
            MState::Init(samples, m) => MState::Init(samples - 1, m.max(sample)),

            MState::Disallow(0, m) => {
                // In the interval QRS ÷ QRS+200ms a new value of M5 is calculated:
                // newM 5 = 0.6*max(Yi)
                let m = 0.6 * m.max(sample) / 5.0; // divide by 5 for averaging

                // The estimated newM 5 value can become quite high, if steep slope premature
                // ventricular contraction or artifact appeared, and for that reason it is
                // limited to newM5 = 1.1* M5 if newM 5 > 1.5* M5.
                let prev_m = self.mm.last().unwrap_or(0.0);
                if m > prev_m * 1.5 {
                    self.mm.push(1.1 * prev_m);
                } else {
                    self.mm.push(m);
                }

                let m = self.m();

                // M is decreased in an interval 225 to 1225 ms [originally 200 to 1200 ms]
                // following the last QRS detection at a low slope, reaching 60 % of its
                // refreshed value at 1225 ms [originally 1200 ms].
                let n_samples = self.fs.s_to_samples(1.0);
                let decrement = m * 0.4 / n_samples as f32;
                self.current_decrement = decrement;

                MState::Decreasing(n_samples, m, decrement)
            }

            // In the interval QRS ÷ QRS+200ms a new value of M5 is calculated:
            // newM 5 = 0.6*max(Yi)
            // Collect maximum value while in Disallow state
            MState::Disallow(samples, m) if sample > m => {
                // if we found a new maximum, extend the disallow period
                MState::Disallow(samples.max(self.fs.s_to_samples(0.2)), sample)
            }
            MState::Disallow(samples, m) => MState::Disallow(samples - 1, m),

            // After 1225 ms [originally 1200 ms] M remains unchanged.
            MState::Decreasing(0, m, _) => MState::ConstantLow(m),

            // M is decreased in an interval 225 to 1225 ms [originally 200 to 1200 ms]
            // following the last QRS detection at a low slope, reaching 60 % of its
            // refreshed value at 1225 ms [originally 1200 ms].
            MState::Decreasing(samples, m, decrease_amount) => {
                // Linear decrease using precomputed decrement value
                MState::Decreasing(samples - 1, m - decrease_amount, decrease_amount)
            }

            // After 1225 ms [originally 1200 ms] M remains unchanged.
            MState::ConstantLow(m) => MState::ConstantLow(m),
        };
    }

    pub fn threshold(&self) -> Option<f32> {
        match self.state {
            MState::Init(_, _) | MState::Disallow(_, _) => None,
            MState::Decreasing(_, m, _) | MState::ConstantLow(m) => Some(m),
        }
    }

    pub fn detection_event(&mut self, sample: f32) {
        // No detection is allowed 225 ms [originally 200 ms] after the current one.
        self.state = MState::Disallow(self.fs.s_to_samples(0.225), sample);
    }
}


src/algorithms/r.rs
use crate::sliding::SlidingWindow;

#[derive(Copy, Clone, Debug)]
enum RState {
    Ignore,
    InitBuffer,
    NoDecrease(u32, u32),    // samples remaining, average rr interval
    Decrease(u32, f32, f32), // samples remaining, value, decrement
    Constant(f32),
}

pub struct R {
    state: RState,
    rr: SlidingWindow<u32, [u32; 5]>,
    prev_idx: u32, // no need to make it an Option
}

impl R {
    pub fn new() -> Self {
        Self {
            state: RState::Ignore,
            rr: SlidingWindow::default(),
            prev_idx: 0,
        }
    }

    pub fn clear(&mut self) {
        self.state = RState::Ignore;
        self.rr.clear();
        self.prev_idx = 0;
    }

    fn enter_no_decrease(&mut self) {
        let rr_sum: u32 = self.rr.iter_unordered().sum();
        let rr_avg = rr_sum / 5;
        self.state = RState::NoDecrease(rr_avg * 2 / 3, rr_avg);
    }

    pub fn update(&mut self, m_decrement: f32) {
        self.state = match self.state {
            RState::NoDecrease(0, rr_avg) => RState::Decrease(rr_avg / 3, 0.0, m_decrement / 1.4),
            RState::NoDecrease(samples, rr_avg) => RState::NoDecrease(samples - 1, rr_avg),
            RState::Decrease(0, r, _) => RState::Constant(r),
            RState::Decrease(samples, r, decrement) => {
                RState::Decrease(samples - 1, r - decrement, decrement)
            }
            o => o,
        };
    }

    pub fn threshold(&self) -> f32 {
        match self.state {
            RState::Ignore | RState::InitBuffer | RState::NoDecrease(_, _) => 0.0,
            RState::Constant(r) | RState::Decrease(_, r, _) => r,
        }
    }

    pub fn detection_event(&mut self, idx: u32) {
        match self.state {
            RState::Ignore => self.state = RState::InitBuffer,
            RState::InitBuffer => {
                self.rr.push(idx.wrapping_sub(self.prev_idx));
                if self.rr.is_full() {
                    self.enter_no_decrease();
                }
            }
            _ => {
                self.rr.push(idx.wrapping_sub(self.prev_idx));
                self.enter_no_decrease();
            }
        };

        self.prev_idx = idx;
    }
}


src/lib.rs
//! This crate provides a realtime ECG QRS detector.
//!
//! The implementation is based on [this article](https://biomedical-engineering-online.biomedcentral.com/articles/10.1186/1475-925X-3-28).
#![cfg_attr(not(test), no_std)]

#[cfg(feature = "alloc")]
extern crate alloc;

mod algorithms;
pub mod sampling;
mod sliding;

use algorithms::{F, M, R};
use sampling::SamplingFrequency;

use crate::sliding::SlidingWindow;

/// Finds QRS complexes in real-time sampled ECG signal.
///
/// # Type parameters:
///
/// - `FMW` - a buffer type containing 300ms worth of samples
/// - `FB` - a buffer type containing 50ms worth of samples
///
/// These type parameters are checked at runtime and, if incorrect, the error message will contain
/// the correct sizes.
pub struct QrsDetector<FMW, FB> {
    total_samples: u32,
    m: M,
    f: F<FMW, FB>,
    r: R,
}

impl QrsDetector<(), ()> {
    /// Creates a new QRS detector for signals sampled with `fs`. The internal buffers will be
    /// allocated on the stack as part of the `QrsDetector` structure.
    ///
    /// # Arguments
    /// * `fs` - The sampling frequency of the processed signal. For more information see
    /// [`sampling::SamplingFrequencyExt`].
    ///
    /// # Example
    /// ```rust
    /// use qrs_detector::sampling::*;
    /// use qrs_detector::QrsDetector;
    ///
    /// // Assuming 500 samples per second
    /// // Type parameters must be 300ms and 50ms in number of samples
    /// let detector = QrsDetector::new::<150, 25>(500.sps());
    /// ```
    pub fn new<const SAMPLES_300: usize, const SAMPLES_50: usize>(
        fs: SamplingFrequency,
    ) -> QrsDetector<[f32; SAMPLES_300], [f32; SAMPLES_50]> {
        QrsDetector {
            total_samples: 0,
            m: M::new(fs),
            f: F::new(fs, SlidingWindow::default(), SlidingWindow::default()),
            r: R::new(),
        }
    }

    /// Creates a new QRS detector for signals sampled with `fs`, using the provided buffers.
    ///
    /// # Arguments
    /// * `fs` - The sampling frequency of the processed signal. For more information see
    /// [`sampling::SamplingFrequencyExt`].
    /// * `f_buffer_300` - A buffer containing 300ms worth of samples.
    /// * `f_buffer_50` - A buffer containing 50ms worth of samples.
    ///
    /// # Examples
    ///
    /// The backing buffers may be slices:
    ///
    /// ```rust
    /// use qrs_detector::sampling::*;
    /// use qrs_detector::QrsDetector;
    ///
    /// // Assuming 500 samples per second
    /// // Buffers must hold 300ms and 50ms worth of samples.
    /// let mut f_buffer_300 = [0.0; 150];
    /// let mut f_buffer_50 = [0.0; 25];
    /// let detector = QrsDetector::new_from(500.sps(), &mut f_buffer_300, &mut f_buffer_50);
    /// ```
    ///
    /// The backing buffers may be arrays:
    ///
    /// ```rust
    /// use qrs_detector::sampling::*;
    /// use qrs_detector::QrsDetector;
    ///
    /// // Assuming 500 samples per second
    /// // Buffers must hold 300ms and 50ms worth of samples.
    /// let f_buffer_300 = [0.0; 150];
    /// let f_buffer_50 = [0.0; 25];
    /// let detector = QrsDetector::new_from(500.sps(), f_buffer_300, f_buffer_50);
    /// ```
    pub fn new_from<FMW, FB>(
        fs: SamplingFrequency,
        f_buffer_300: FMW,
        f_buffer_50: FB,
    ) -> QrsDetector<FMW, FB>
    where
        FMW: AsRef<[f32]> + AsMut<[f32]>,
        FB: AsRef<[f32]> + AsMut<[f32]>,
    {
        QrsDetector {
            total_samples: 0,
            m: M::new(fs),
            f: F::new(
                fs,
                SlidingWindow::new(f_buffer_300),
                SlidingWindow::new(f_buffer_50),
            ),
            r: R::new(),
        }
    }

    /// Creates a new QRS detector for signals sampled with `fs`, using the provided buffers.
    ///
    /// # Arguments
    /// * `fs` - The sampling frequency of the processed signal. For more information see
    /// [`sampling::SamplingFrequencyExt`].
    /// * `f_buffer_300` - A buffer containing 300ms worth of samples.
    /// * `f_buffer_50` - A buffer containing 50ms worth of samples.
    ///
    /// # Examples
    ///
    /// The backing buffers may be slices:
    ///
    /// ```rust
    /// use qrs_detector::sampling::*;
    /// use qrs_detector::QrsDetector;
    ///
    /// // Assuming 500 samples per second
    /// // Buffers must hold 300ms and 50ms worth of samples.
    /// let mut f_buffer_300 = [0.0; 150];
    /// let mut f_buffer_50 = [0.0; 25];
    /// let detector = QrsDetector::new_from(500.sps(), &mut f_buffer_300, &mut f_buffer_50);
    /// ```
    ///
    /// The backing buffers may be arrays:
    ///
    /// ```rust
    /// use qrs_detector::sampling::*;
    /// use qrs_detector::QrsDetector;
    ///
    /// // Assuming 500 samples per second
    /// // Buffers must hold 300ms and 50ms worth of samples.
    /// let f_buffer_300 = [0.0; 150];
    /// let f_buffer_50 = [0.0; 25];
    /// let detector = QrsDetector::new_from(500.sps(), f_buffer_300, f_buffer_50);
    /// ```
    #[cfg(feature = "alloc")]
    pub fn new_alloc(
        fs: SamplingFrequency,
    ) -> QrsDetector<alloc::boxed::Box<[f32]>, alloc::boxed::Box<[f32]>> {
        use alloc::vec;
        QrsDetector {
            total_samples: 0,
            m: M::new(fs),
            f: F::new(
                fs,
                SlidingWindow::new(vec![0.0; fs.ms_to_samples(300.0)].into_boxed_slice()),
                SlidingWindow::new(vec![0.0; fs.ms_to_samples(50.0)].into_boxed_slice()),
            ),
            r: R::new(),
        }
    }
}

impl<FMW, FB> QrsDetector<FMW, FB>
where
    FMW: AsRef<[f32]> + AsMut<[f32]>,
    FB: AsRef<[f32]> + AsMut<[f32]>,
{
    /// Resets the internal state of the detector.
    pub fn clear(&mut self) {
        self.m.clear();
        self.f.clear();
        self.r.clear();
    }

    /// Processes a sample. Returns Some sample index if a QRS complex is detected.
    pub fn update(&mut self, sample: f32) -> Option<u32> {
        self.m.update(sample);
        self.f.update(sample);
        self.r.update(self.m.current_decrement);

        let thresholds = self.thresholds();

        let result = match thresholds.total() {
            Some(mfr) if sample > mfr => {
                self.m.detection_event(sample);
                self.r.detection_event(self.total_samples);
                Some(self.total_samples)
            }
            _ => None,
        };

        self.total_samples += 1;
        result
    }

    /// Returns the current threshold value.
    /// This value is used to determine if a sample is a QRS complex.
    /// The final threshold is calculated as `M + F + R`.
    pub fn thresholds(&self) -> Thresholds {
        Thresholds {
            m: self.m.threshold(),
            f: self.f.threshold(),
            r: self.r.threshold(),
        }
    }
}

#[derive(Clone, Copy, Debug)]
pub struct Thresholds {
    pub m: Option<f32>,
    pub f: Option<f32>,
    pub r: f32,
}

impl Thresholds {
    pub fn total(&self) -> Option<f32> {
        if let (Some(m), Some(f)) = (self.m, self.f) {
            Some(m + f + self.r)
        } else {
            None
        }
    }
}
